### DISCUSSION TITLE
Confused about the ability to run multiple tests

### DISCUSSION THREAD

**Discussion by 2dm** [esnet/iperf#1312] [Q&A] [answered]
Hi,

There are a few conflicting resources about whether I can use iperf3 to run simultaneous tests.
For example, [this answer](https://github.com/esnet/iperf/discussions/1276#discussioncomment-2311868) and on the other hand [this page](https://fasterdata.es.net/performance-testing/network-troubleshooting-tools/iperf/multi-stream-iperf3/) (which I reached from the FAQ page).

Is it possible to use multiple, separated by port, iperf3 processes to run simultaneous test? 
If not, I would be glad to hear some details about what causes it to be not reliable. (since results look reasonable if I try it)
Thank you.

**Comment by davidBar-On** [comment]
Hi, it is possible to run several servers using different ports.  It may be that part is your confusion is because it depends where is the throughput limit.  For example, when several servers run on a machine with high throughput and the clients run from different mobile devices with relatively low throughput, then all parallel test will get the maximum throughput allowed from the mobiles.  On the other hand, if all clients run from the same machine, or the server cannot support the accumulated throughput from all clients, then running several tests in parallel may not give the right results.

**Reply by 2dm** [reply]
@davidBar-On, thank you for answering!

What do you mean by "cannot support the accumulated throughput"? Because of CPU limitation? 
For example, if I want to run 10 connections from one client to one server through a congested link of 1gpbs (cpu is not a bottleneck in this example). I expect each connection will have average throughput of 100mbps, given fairness, same priority etc.
Where is the "point of failure" that may produce wrong results?

**Comment by davidBar-On** [comment, marked as answer]
> What do you mean by "cannot support the accumulated throughput"? Because of CPU limitation?

Bottleneck can be either CPU link capacity.  I mainly meant the link capacity.  Your example test case may work as you described, but the problem is that it doesn't measure the maximum throughput from each client.  E.g. if each client can send 200mbps this will not be shown in the test you described.  Of course, the test is good it represents the use case your are evaluating.

**Comment by 2dm** [comment]
OK, I think I now understand what you mean. 
If I try to summaries it: 

1. Multiple simultaneous iperfs _can_ run together when I wish to test how a client, or all clients, perform in a certain network scenario.
2. Different iperf processes which are separated by port (e.g. client 1 and client 2 on a single host, and same for servers) _do not_
 interfere with each other, other than CPU usage and traffic that shares interfaces and links. 
3. To evaluate the _maximum_ performance of a client under a given topology, which is the main purpose of iperf3, run a single client and a single server - unless link capacity is high (lets say >25Gbps) where a single iperf may not be enough to achieve full utilization.
Is that correct?

I think my confusion was that I understand from other discussions/issues answers that "incorrect test results" means that the output does not represents what really happened to its process. 
Thanks @davidBar-On!

**Comment by davidBar-On** [comment]
@2dm, regarding 3, this is correct.  Note that the the link capacity depends on the full path between the client and the server.  For example, in many cases, a client on a mobile device maximum throughput is much lower then the link to the server, because of the limited capacity of the cellular network.  Therefore, several mobile clients may be able to connect to the same server machine (running several iper3 servers) without any throughput degradation.
